### PR TITLE
[core] Post `_on_completed` callbacks to `io_service_` instead of running them synchronously

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Attempts to fix the deadlock described in the linked issue by posting callbacks to the `io_service_`. This avoids running them while holding locks (such as from the task manager).

## Related issue number

https://github.com/ray-project/ray/issues/39113 (will be closed once cherry-picked).

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
Kicked off release tests for core and serve: https://buildkite.com/ray-project/release-tests-pr/builds/51462

**Comment by rkooo567** [comment]
Test test seems pretty promising

**Comment by rkooo567** [comment]
previous release tests are here; https://buildkite.com/ray-project/release-tests-pr/builds/51486

I am rerunning core tests

**Comment by rkooo567** [comment]
@edoakes can you take a look at https://buildkite.com/ray-project/release-tests-pr/builds/51486#018a485f-94f3-4afa-96a5-3f3ea7cab789?

**Comment by rkooo567** [comment]
Rerunning it with core https://buildkite.com/ray-project/release-tests-pr/builds/51571. I hope the wheel is correct this time...

**Comment by rkooo567** [comment]
Test result so far seems pretty promising

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -3993,23 +3993,40 @@ void CoreWorker::YieldCurrentFiber(FiberEvent &event) {
...
-                              fallback_callback(ray_object, object_id, python_future);
-                            } else {
-                              success_callback(ray_object, object_id, python_future);
-                            }
-                          });
+  memory_store_->GetAsync(
+      object_id,
+      [this,
+       object_id,
+       success_callback,
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:4008]:
```suggestion
       success_callback=std::move(success_callback),
```

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -3993,23 +3993,40 @@ void CoreWorker::YieldCurrentFiber(FiberEvent &event) {
...
+       python_user_callback,
+       fallback_callback =
+           std::move(fallback_callback)](std::shared_ptr<RayObject> ray_object) {
+        // Post the callback to the io_service_ to avoid deadlocks.
+        // The user callback can make arbitrary Ray API calls and will be called
+        // immediately when the object is `Put` into the in-memory store. This can
+        // cause deadlocks if the callers of `Put` is holding a lock.
+        io_service_.post(
+            [object_id,
+             success_callback,
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:4018]:
```suggestion
             success_callback=std::move(success_callback),
```

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -3993,23 +3993,40 @@ void CoreWorker::YieldCurrentFiber(FiberEvent &event) {
...
+             python_user_callback,
+             fallback_callback = std::move(fallback_callback),
+             ray_object = std::move(ray_object)]() {
+              if (ray_object->IsInPlasmaError()) {
+                fallback_callback(ray_object, object_id, python_user_callback);
+              } else {
+                success_callback(ray_object, object_id, python_user_callback);
+              }
+            },
+            "CoreWorker.GetAsync callback");
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:4028]: Let's follow the convention from other parts? 
```suggestion
            "CoreWorker.GetAsync.Callback");
```

**Review by rkooo567 — APPROVED** [review]
LGTM. The test result seems promising

**Comment by edoakes** [comment]
@rkooo567 there are a lot of failures that have tracebacks like this:
```
28338 ^[[2m^[[36m(pid=149561)^[[0m Traceback (most recent call last):
28339 ^[[2m^[[36m(pid=149561)^[[0m   File "python/ray/_raylet.pyx", line 1997, in ray._raylet.task_execution_handler
28340 ^[[2m^[[36m(pid=149561)^[[0m   File "python/ray/_raylet.pyx", line 1853, in ray._raylet.execute_task_with_cancellation_handler
28341 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/function_manager.py", line 574, in load_actor_class
28342 ^[[2m^[[36m(pid=149561)^[[0m     actor_class = self._load_actor_class_from_gcs(
28343 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/function_manager.py", line 677, in _load_actor_class_from_gcs
28344 ^[[2m^[[36m(pid=149561)^[[0m     actor_class = pickle.loads(pickled_class)
28345 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/serve/__init__.py", line 4, in <module>
28346 ^[[2m^[[36m(pid=149561)^[[0m     from ray.serve.api import (
28347 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/serve/api.py", line 15, in <module>
28348 ^[[2m^[[36m(pid=149561)^[[0m     from ray.serve.built_application import BuiltApplication
28349 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/serve/built_application.py", line 7, in <module>
28350 ^[[2m^[[36m(pid=149561)^[[0m     from ray.serve.deployment import Deployment
28351 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/serve/deployment.py", line 24, in <module>
28352 ^[[2m^[[36m(pid=149561)^[[0m     from ray.serve.context import get_global_client
28353 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/serve/context.py", line 12, in <module>
28354 ^[[2m^[[36m(pid=149561)^[[0m     from ray.serve._private.client import ServeControllerClient
28355 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/serve/_private/client.py", line 27, in <module>
28356 ^[[2m^[[36m(pid=149561)^[[0m     from ray.serve._private.deploy_utils import get_deploy_args
28357 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/serve/_private/deploy_utils.py", line 8, in <module>
28358 ^[[2m^[[36m(pid=149561)^[[0m     from ray.serve.schema import ServeApplicationSchema
28359 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/serve/schema.py", line 856, in <module>
28360 ^[[2m^[[36m(pid=149561)^[[0m     class ServeInstanceDetails(BaseModel, extra=Extra.forbid):
28361 ^[[2m^[[36m(pid=149561)^[[0m   File "pydantic/main.py", line 197, in pydantic.main.ModelMetaclass.__new__
28362 ^[[2m^[[36m(pid=149561)^[[0m   File "pydantic/fields.py", line 497, in pydantic.fields.ModelField.infer
28363 ^[[2m^[[36m(pid=149561)^[[0m   File "pydantic/fields.py", line 460, in pydantic.fields.ModelField._get_field_info
28364 ^[[2m^[[36m(pid=149561)^[[0m   File "pydantic/typing.py", line 120, in pydantic.typing.get_origin
28365 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/typing.py", line 1271, in get_origin
28366 ^[[2m^[[36m(pid=149561)^[[0m     def get_origin(tp):
28367 ^[[2m^[[36m(pid=149561)^[[0m   File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/worker.py", line 778, in sigterm_handler
28368 ^[[2m^[[36m(pid=149561)^[[0m     sys.exit(1)
28369 ^[[2m^[[36m(pid=149561)^[[0m SystemExit: 1
28370 ^[[2m^[[36m(pid=149561)^[[0m AttributeError: 'Worker' object has no attribute 'core_worker'
```

Looks sort of suspicious (some others had this raised in `terminate_asyncio_thread` which looks related). Let me check other recent runs to see if this is new.

**Comment by edoakes** [comment]
Same tracebacks are happening on the release branch, so I'm not concerned about them: https://buildkite.com/ray-project/release-tests-branch/builds/2103#018a44b2-6415-495b-8823-7824101415f8

Note that this test also forcibly kills replicas constantly, so seeing some tracebacks is expected.

**Comment by edoakes** [comment]
All tests also passed on the serve cancellation PR with this change merged in: https://buildkite.com/ray-project/oss-ci-build-pr/builds/34426

**Pull request merged**